libraries/AP_PacketDigitalCAN/AP_PacketDigitalCAN.cpp

```diff
@@ -0,0 +1,609 @@
+        }
+    } while (!ret || !write_select);
+
+    uint64_t deadline = AP_HAL::native_micros64() + 2000000;
+    return (_can_iface->send(out_frame, deadline, AP_HAL::CANIface::AbortOnError) == 1);
+}
+
+// read frame on CAN bus, returns true on succses
+bool AP_PacketDigitalCAN::read_frame(AP_HAL::CANFrame &recv_frame, const uint64_t timeout, AP_HAL::CANIface::CanIOFlags &flags)
+{
+    if (!_initialized) {
+        debug_can(AP_CANManager::LOG_ERROR, "Driver not initialized for read_frame\n\r");
+        return false;
+    }
+    // wait for space in buffer to read
+    bool read_select = true;
+    bool write_select = false;
+    bool ret = _can_iface->select(read_select, write_select, nullptr, AP_HAL::native_micros64() + timeout);
+    if (!ret || !read_select) {
+        // return false if no data is available to read
+        return false;
+    }
+
+    uint64_t time;
+
+    // read frame and return success
+    return (_can_iface->receive(recv_frame, time, flags) == 1);
+}
+
+bool AP_PacketDigitalCAN::pre_arm_check(char* reason, uint8_t reason_len) const
+{
+    for (uint8_t i=0; i<device_count; i++) {
+        if (!is_healthy_by_index(i)) {
+            return false;
+        }
+    }
+    return true;
+}
+
+
+// return fault code as string
+const char* AP_PacketDigitalCAN::get_fault_code_string(const Fault_Flags fault) const
+{
+    switch (fault) {
+    case Fault_Flags::NONE:
+        return "None";
+    case Fault_Flags::OVER_VOLTAGE:
+        return "Over-Voltage";
+    case Fault_Flags::UNDER_VOLTAGE:
+        return "Under-Voltage";
+    case Fault_Flags::OVER_CURRENT:
+        return "Over-Current";
+    case Fault_Flags::OVER_TEMPERATURE:
+        return "Over-Temperature";
+    default:
+        return "Unknown";
+    }
+}
+
+bool AP_PacketDigitalCAN::get_voltage_and_current_and_temp(const uint16_t serialnumber, const bool chkInput, float &voltage, float &current, float &temperature) const
+{
+    if (device_count == 0) {
+        return false;
+    }
+
+    if (serialnumber == 0) {
+        // take the average output of all healthy devices
+        int32_t count = 0;
+        float voltage_out_avg = 0.0f;
+        float current_out_avg = 0.0f;
+        float temperature_avg = 0.0f;
+
+        for (uint8_t i=0; i<device_count; i++) {
+            if (!MPPT_devices[i].is_healthy()) {
+                continue;
+            }
+            count++;
+            voltage_out_avg += MPPT_devices[i].output.voltage;
+            current_out_avg += MPPT_devices[i].output.current;
+            temperature_avg += MPPT_devices[i].temperature;
+        }
+        if (count > 0) {
+            voltage = voltage_out_avg / count;
+            current = current_out_avg / count;
+            temperature = (float)temperature_avg / count;
+            // average OUTPUTs of all healthy devices
+            return true;
+        }
+        // no healthy devices found
+        return false;
+
+    } else {
```

Comment:
Don't need an else; the average block returns.

Reply:
removed